### PR TITLE
Request kube apiserver directly from frontend

### PR DESCRIPTION
**Pull request by momom-i**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
To request directry from frontend

#### Which issue(s) this PR fixes:

Fixes https://github.com/kubernetes-sigs/kube-scheduler-simulator/issues/95

#### Special notes for your reviewer:

@sanposhiho Thank you for answering my questions so many times.
After some discussion, we decided to fix the kube-apiserver URL and throw resource change requests directly to that URL.

I tried a few things and confirmed that it returned same results as previous implementation.
Please let me know if there is any part of the behavior that is different from the previous one.

Also I'd like to learn more so correcting my every little mistake is welcome!

/label tide/merge-method-squash

### CODE REVIEW

File: k8sapiserver/k8sapiserver.go [hunk 1]
```diff
@@ -130,6 +130,8 @@ func NewControlPlaneConfigWithOptions(serverURL, etcdURL string) *controlplane.C
 
 	genericConfig.SecureServing = &genericapiserver.SecureServingInfo{Listener: fakeLocalhost443Listener{}}
 
+	genericConfig.CorsAllowedOriginList = []string{"http://*"}
```
Comment (196Ikuchil) [inline k8sapiserver/k8sapiserver.go:133]: Hi @momom-i!
Cc. @sanposhiho 

From a perspective of security, you should indicate the frontend url like this, I think.
https://github.com/kubernetes-sigs/kube-scheduler-simulator/blob/fc420bc2e9e55fedc13adba583ee41c9020387fa/server/server.go#L31
(And current code  maybe not  allow a request from `HTTPS`.)

But this `genericConfig` is also applied to `ectd`. Unfortunately, the above configuration is for the `API server`. (The `etcd` is accessed by only`API server`.)
 Therefore , you should separate that configuration logic I think.

Comment (sanposhiho) [reply k8sapiserver/k8sapiserver.go:133]: > But this genericConfig is also applied to ectd. Unfortunately, the above configuration is for the API server. (The etcd is accessed by onlyAPI server.)

@196Ikuchil Is there any documentation for this? I did not know about this. (Either way, I think @196Ikuchil's idea is valid :) )

**Comment by sanposhiho** [comment]
@momom-i Thanks for a great work. I'll review this in this weekend.

File: k8sapiserver/k8sapiserver.go (hunk 1 above)
Comment (196Ikuchil) [reply k8sapiserver/k8sapiserver.go:133]: > But this genericConfig is also applied to ectd.

The point I was trying to make is that `genericConfig` is used in both 
[etcdOptions.ApplyWithStorageFactoryTo...](https://github.com/kubernetes-sigs/kube-scheduler-simulator/pull/115/files/d63a12c1166b215ac156f1e7c05b774334d329a7#diff-7dc54cbad11a8158cbb340b7757b85aef66049765a514b15908cf940a6d44185R134) and [cfg := &controlplane.Config{GenericConfig: genericConfig,...](https://github.com/kubernetes-sigs/kube-scheduler-simulator/pull/115/files/d63a12c1166b215ac156f1e7c05b774334d329a7#diff-7dc54cbad11a8158cbb340b7757b85aef66049765a514b15908cf940a6d44185R140)
And I think it is not good to apply that `AllowOrigins` configuration to `etcd`.

Therefore, there is no documentation, sorry.

Comment (sanposhiho) [reply k8sapiserver/k8sapiserver.go:133]: I had a quick look at the implementation of `ApplyWithStorageFactoryTo`. But I don't think the AllowOrigins configuration is applying to etcd. (In the first place, there is no `AllowOrigins` field in the etcd options.)

@momom-i 
So, using `genericConfig. CorsAllowedOriginList` looks good to me, please change `"http://*"` to use `cfg.FrontendURL` like the simulator server.
https://github.com/kubernetes-sigs/kube-scheduler-simulator/blob/fc420bc2e9e55fedc13adba583ee41c9020387fa/server/server.go#L31

Comment (momom-i) [reply k8sapiserver/k8sapiserver.go:133]: I'm sorry for my late reply. Thank you for your comment, @196Ikuchil and also @sanposhiho !
I've confirmed `EtcdOptions` has no `AllowOrigins` field,  so I'll change like what you said.

File: k8sapiserver/k8sapiserver.go [hunk 2]
```diff
@@ -130,6 +130,8 @@ func NewControlPlaneConfigWithOptions(serverURL, etcdURL string) *controlplane.C
 
 	genericConfig.SecureServing = &genericapiserver.SecureServingInfo{Listener: fakeLocalhost443Listener{}}
 
+	genericConfig.CorsAllowedOriginList = []string{frontendURL}
```
Comment (sanposhiho) [inline k8sapiserver/k8sapiserver.go:133]: 👍

**Review by sanposhiho — COMMENTED** [review]
Change the default KUBE_API_PORT value to 3131.
https://github.com/kubernetes-sigs/kube-scheduler-simulator/blob/master/config/config.go#L84

File: config/config.go [hunk 3]
```diff
@@ -81,7 +81,7 @@ func getKubeAPIServerURL() string {
 	if p == "" {
 		// we still want the simulator to behave as before,
 		// use a local test port.
```
Comment (sanposhiho) [inline config/config.go:83]: Remove this line.

File: web/api/v1/node.ts [hunk 4]
```diff
@@ -1,26 +1,66 @@
...
+      { headers: { "Content-Type": "application/strategic-merge-patch+json" } }
+    );
     return res.data;
   } catch (e: any) {
-    onError(e);
+    if (axios.isAxiosError(e) && e.response && e.response.status === 404) {
+      const res = await createNode(req, onError);
+      return res;
+    }
+    onError("Caused by applyNode: " + e);
```
Comment (sanposhiho) [inline web/api/v1/node.ts:22]: Please change other wordings as well.

```suggestion
    onError("failed to apply Node: " + e);
```

**Review by sanposhiho — CHANGES_REQUESTED** [review]

File: config/config.go (hunk 3 above)
Comment (momom-i) [reply config/config.go]: fixed: [f4deb1d](https://github.com/kubernetes-sigs/kube-scheduler-simulator/pull/115/commits/f4deb1d59c6dbe294e4d9f4ddd27f61ca990dab0)

File: web/api/v1/node.ts (hunk 4 above)
Comment (momom-i) [reply web/api/v1/node.ts]: fixed: 3f20091

File: web/api/v1/pod.ts [hunk 5]
```diff
@@ -1,26 +1,69 @@
...
+    }
+    const res = await k8sInstance.patch<V1Pod>(
+      namespaceURL + `/pods/${req.metadata.name}?fieldManager=simulator`,
+      req,
+      { headers: { "Content-Type": "application/strategic-merge-patch+json" } }
+    );
     return res.data;
   } catch (e: any) {
-    onError(e);
+    if (axios.isAxiosError(e) && e.response && e.response.status === 404) {
```
Comment (sanposhiho) [inline web/api/v1/pod.ts:18]: I think this 404 error should be resolved by using server side apply. You currently use patch.
So, use `Content-Type` to `application/apply-patch+yaml` instead.
https://kubernetes.io/docs/reference/using-api/server-side-apply/#api-endpoint

Comment (sanposhiho) [reply web/api/v1/pod.ts:18]: And if the error is gone, you can delete `createPod` :)

Comment (momom-i) [reply web/api/v1/pod.ts:18]: Thank you, this opportunity gave me learn and understand server side apply!

**Review by sanposhiho — COMMENTED** [review]
When we remove a Node, we also need to remove Pod(s) on that Node.  (like backend does: [ref](https://github.com/kubernetes-sigs/kube-scheduler-simulator/blob/master/node/node.go#L77))
You can see why we need to remove the Pod(s) with the below example:
1. create Node1
2. create Pod1 (and Pod1 goes to Node1)
3.  remove Node1
4.  create Node1 again
5.  Pod1 is still on Node1

So, could you change this `deleteOnClick` to remove Pods on the Node when remove the Node?
https://github.com/kubernetes-sigs/kube-scheduler-simulator/blob/master/web/components/ResourceBar/ResourceBar.vue#L240

**Comment by 196Ikuchil** [comment]
Hi @momom-i, 
these conflicts are mainly occurred by PR #119.

In that PR, I created a new directory `web/components/ResourceViews`, new file `ResourceViewPanel.vue` and moved some `XXXList.vue` to under the `web/components/ResourceViews/Lists` directory.
So, you should check this [web/pages/index.vue](https://github.com/kubernetes-sigs/kube-scheduler-simulator/pull/119/files#diff-d69e128299ebce389ac7f8eb7d5d2b294e9c600cf871cae4c3571db88511553e) file first, please?

If you have any questions, ask me, please.

**Comment by 196Ikuchil** [comment]
Thank you! I checked the ListViews on my PC and it looks good maybe.

Btw, I get a response `400 Bad request` when I have tried to `Apply` button in WebUI. 
If you understand the cause of the error, please fix it? (sorry I didn't check it in detail)

The following is an example of sending an `apply` request to update a Node's `metadata.labels` value.
### Check in master branch
```
POST /api/v1/nodes HTTP/1.1
Host: localhost:1212
User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:97.0) Gecko/20100101 Firefox/97.0
Accept: application/json, text/plain, */*
Accept-Language: ja,en-US;q=0.7,en;q=0.3
Accept-Encoding: gzip, deflate
Content-Type: application/json;charset=utf-8
Content-Length: 866
Origin: http://localhost:3000
Connection: close
Referer: http://localhost:3000/
Cookie: _ga_F1R581XGNQ=GS1.1.1635005988.2.0.1635005988.0; _ga=GA1.1.1515817854.1635000707; __gads=ID=6aa81bf2479372d8-228a3956d9cc0053:T=1635000710:RT=1635000710:S=ALNI_MZd4pR1Dkc5FmAZmSASTvT9zXLx9g
Sec-Fetch-Dest: empty
Sec-Fetch-Mode: cors
Sec-Fetch-Site: same-site

{"metadata":{"name":"node2","uid":"0be1e7ac-0386-446d-bac2-ae277c4ce621","resourceVersion":"182","creationTimestamp":"2022-03-17T15:26:00Z","labels":{"test":"hoge-override"},"managedFields":[{"manager":"simulator","operation":"Apply","apiVersion":"v1","time":"2022-03-17T15:26:00Z","fieldsType":"FieldsV1","fieldsV1":{"f:metadata":{"f:labels":{"f:test":{}}},"f:spec":{}}}]},"spec":{},"status":{"capacity":{"cpu":"4","memory":"32Gi","pods":"110"},"allocatable":{"cpu":"4","memory":"32Gi","pods":"110"},"phase":"Running","conditions":[{"type":"Ready","status":"True","lastHeartbeatTime":null,"lastTransitionTime":null}],"daemonEndpoints":{"kubeletEndpoint":{"Port":0}},"nodeInfo":{"machineID":"","systemUUID":"","bootID":"","kernelVersion":"","osImage":"","containerRuntimeVersion":"","kubeletVersion":"","kubeProxyVersion":"","operatingSystem":"","architecture":""}}}


HTTP/1.1 200 OK
Access-Control-Allow-Credentials: true
Access-Control-Allow-Origin: http://localhost:3000
Content-Type: application/json; charset=UTF-8
Vary: Origin
Date: Thu, 17 Mar 2022 15:26:27 GMT
Content-Length: 867
Connection: close

{"metadata":{"name":"node2","uid":"0be1e7ac-0386-446d-bac2-ae277c4ce621","resourceVersion":"183","creationTimestamp":"2022-03-17T15:26:00Z","labels":{"test":"hoge-override"},"managedFields":[{"manager":"simulator","operation":"Apply","apiVersion":"v1","time":"2022-03-17T15:26:27Z","fieldsType":"FieldsV1","fieldsV1":{"f:metadata":{"f:labels":{"f:test":{}}},"f:spec":{}}}]},"spec":{},"status":{"capacity":{"cpu":"4","memory":"32Gi","pods":"110"},"allocatable":{"cpu":"4","memory":"32Gi","pods":"110"},"phase":"Running","conditions":[{"type":"Ready","status":"True","lastHeartbeatTime":null,"lastTransitionTime":null}],"daemonEndpoints":{"kubeletEndpoint":{"Port":0}},"nodeInfo":{"machineID":"","systemUUID":"","bootID":"","kernelVersion":"","osImage":"","containerRuntimeVersion":"","kubeletVersion":"","kubeProxyVersion":"","operatingSystem":"","architecture":""}}}
```

### Check in this PR

```
PATCH /api/v1/nodes/node1?fieldManager=simulator HTTP/1.1
Host: localhost:3131
User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:97.0) Gecko/20100101 Firefox/97.0
Accept: application/json, text/plain, */*
Accept-Language: ja,en-US;q=0.7,en;q=0.3
Accept-Encoding: gzip, deflate
Content-Type: application/apply-patch+yaml
Content-Length: 898
Origin: http://localhost:3000
Connection: close
Referer: http://localhost:3000/
Cookie: _ga_F1R581XGNQ=GS1.1.1635005988.2.0.1635005988.0; _ga=GA1.1.1515817854.1635000707; __gads=ID=6aa81bf2479372d8-228a3956d9cc0053:T=1635000710:RT=1635000710:S=ALNI_MZd4pR1Dkc5FmAZmSASTvT9zXLx9g
Sec-Fetch-Dest: empty
Sec-Fetch-Mode: cors
Sec-Fetch-Site: same-site

{"kind":"Node","apiVersion":"v1","metadata":{"name":"node1","uid":"c0046bf6-ab5c-4b2e-b48b-7aae527c984f","resourceVersion":"178","creationTimestamp":"2022-03-17T15:30:42Z","labels":{"test":"hoge-override"},"managedFields":[{"manager":"simulator","operation":"Apply","apiVersion":"v1","time":"2022-03-17T15:30:42Z","fieldsType":"FieldsV1","fieldsV1":{"f:metadata":{"f:labels":{"f:test":{}}},"f:spec":{}}}]},"spec":{},"status":{"capacity":{"cpu":"4","memory":"32Gi","pods":"110"},"allocatable":{"cpu":"4","memory":"32Gi","pods":"110"},"phase":"Running","conditions":[{"type":"Ready","status":"True","lastHeartbeatTime":null,"lastTransitionTime":null}],"daemonEndpoints":{"kubeletEndpoint":{"Port":0}},"nodeInfo":{"machineID":"","systemUUID":"","bootID":"","kernelVersion":"","osImage":"","containerRuntimeVersion":"","kubeletVersion":"","kubeProxyVersion":"","operatingSystem":"","architecture":""}}}


HTTP/1.1 400 Bad Request
Access-Control-Allow-Credentials: true
Access-Control-Allow-Headers: Content-Type, Content-Length, Accept-Encoding, X-CSRF-Token, Authorization, X-Requested-With, If-Modified-Since
Access-Control-Allow-Methods: POST, GET, OPTIONS, PUT, DELETE, PATCH
Access-Control-Allow-Origin: http://localhost:3000
Access-Control-Expose-Headers: Date
Audit-Id: eb725e0f-5e05-4890-8bf1-099864908b5e
Cache-Control: no-cache, private
Content-Type: application/json
X-Kubernetes-Pf-Flowschema-Uid: 1f59e016-5310-4ede-84f5-4dd66b67f6c1
X-Kubernetes-Pf-Prioritylevel-Uid: cbee2024-0a45-497f-b53b-cabceeaff462
Date: Thu, 17 Mar 2022 15:30:56 GMT
Content-Length: 185
Connection: close

{
  "kind": "Status",
  "apiVersion": "v1",
  "metadata": {
    
  },
  "status": "Failure",
  "message": "metadata.managedFields must be nil",
  "reason": "BadRequest",
  "code": 400
}
```

**Comment by momom-i** [comment]
Thank you for your check, @196Ikuchil!
I've confirmed it occurred on my local too. It seems caused by changing to server-side apply.
I'll change to remove `metadata.managedFields` before updating resources later.

File: web/components/ResourceBar/ResourceBar.vue [hunk 6]
```diff
@@ -232,17 +232,33 @@ export default defineComponent({
...
       if (store) {
         const y = yaml.load(formData.value);
-        store.apply(y, setServerErrorMessage);
+        store.apply(y).catch((e) => setServerErrorMessage(e));
       }
       drawer.value = false;
     };
 
     const deleteOnClick = () => {
+      if (selected.value?.resourceKind === "Node") {
```
Comment (sanposhiho) [inline web/components/ResourceBar/ResourceBar.vue:241]:
```suggestion
      if (selected.value?.resourceKind === "Node") {
      // when the Node is deleted, all Pods on the Node should be deleted as well.
```

**Review by sanposhiho — COMMENTED** [review]
just one nits

**Review by sanposhiho — COMMENTED** [review]
Thanks for a great work!!!
/lgtm
/approve

**Pull request merged**